Update frontend for category descriptions and user auth

seerocode opened the pull request:
### Changes for categoryvalue update from backend:
- Update mock resolvers for failing tests
- Update dataset records table and entry form render function

seerocode committed:
ch247(chore): update client schema

seerocode committed:
chore(ch247): update client mutations and queries for category value

seerocode committed:
ch247(chore): update data entry form and records table for category value

seerocode committed:
chore(ch247): update mock resolvers

seerocode committed:
refactor(ch247): set inputter id in entry from api context
- inputter id is not required in schema because it is set in the entry on record create and update

seerocode committed:
fix: add auth middleware with temp header

seerocode committed:
fix: add missing session param

seerocode committed:
fix: add permissions mixin to target

seerocode commented on api/schema.graphql:
```diff
@@ -200,7 +200,7 @@ input UpdateCategoryValueInput {
 input EntryInput {
   id: ID
-  inputterId: ID!
+  inputterId: ID
```
We might not need this here anymore but leaving it be for now.

jnu replied:
I think we should actually remove this. currently it's a security problem to leave it in -- we'd want to validate that the `current_user` matches the `inputterId` anyway, which we don't currently do. (otherwise you could enter data as someone else.)

there's a tiny possibility that someone will want permission to enter on behalf of someone else... i don't think we should implement that unless we absolutely have to!

seerocode replied:
I removed the field in the schema and updated the record tests. CC @nmanduj1

seerocode commented on client/src/index.tsx:
```diff
@@ -30,6 +31,17 @@ const httpLink = new HttpLink({
+  // Use the setContext method to set the HTTP headers
+  operation.setContext({
+    headers: {
+      "X-User": "tester@notrealemail.info", // TODO: remove placeholder when user login is merged
```
Pending merge of login UI work to remove the placeholder.

jnu replied:
glad you got this to work around! if development with a login becomes tedious we can discuss more workarounds to make it easier.

seerocode commented on api/mutations.py:
```diff
@@ -88,10 +88,12 @@ def resolve_create_record(obj, info, input):
     '''
 
     session = info.context['dbsession']
+    current_user = info.context['current_user']
```
I moved the inputter id setting to the resolver since there isn't a way to get this info right now from the frontend. Also referencing the comment here: https://github.com/stanford-policylab/bbc-50-50/pull/68#discussion_r651831754

jnu replied:
this makes sense to me! i think this is the right way to go regardless of whether we can get the right ID in the UI or not.

jnu reviewed (approved):
nice work!

seerocode committed:
feat: add error component
- used in home page

seerocode committed:
feat: add user email to apollo header and auth id to user query
- add functions to get user id and email from user profile in auth provider
- use auth hook to get user id for user queries in sidebar and homepage
- update auth link for apollo client to get user email for x-user header from auth service

seerocode committed:
test: update tests to include auth mock
> - sidebar, home updated

seerocode committed:
chore: remove inputter id from entry input

seerocode committed:
test: fix inputter id result for create record test

seerocode commented on api/app_test.py:
```diff
@@ -782,12 +782,12 @@ def test_create_record(self):
                 },
             }, user=user)
 
+            # self.maxDiff = None
```
I'll be removing this in a bit.

jnu replied:
it's fine with me if you want to keep the maxDiff setting in place, i'm only _not_ a fan of commented out code lines :) 

you can make it an attribute of the class if you want so it applies to all the test cases:

```py
class TestGraphQL(unittest.TestCase):

    maxDiff = None

    def setUp():
        # etc...
```

seerocode commented on api/schema.graphql:
```diff
@@ -200,7 +200,6 @@ input UpdateCategoryValueInput {
 
 input EntryInput {
   id: ID
-  inputterId: ID!
```
I removed inputterId from the schema to resolve the possible security issue and we are setting this in the resolver.

seerocode commented on client/src/components/Home/Home.tsx:
```diff
@@ -102,16 +104,19 @@ const getTableData = (
     GET_USER,
     {
-      variables: { id: "cd7e6d44-4b4d-4d7a-8a67-31efffe53e77" }, // TODO: Replace after user auth implementation
+      variables: { id: userId },
```
This is getting the user id from the currently logged in user.

seerocode commented on client/src/index.tsx:
```diff
@@ -33,6 +34,17 @@ const httpLink = new HttpLink({
+  // Use the setContext method to set the HTTP headers
+  operation.setContext(({ headers = {} }) => ({
+    headers: {
+      "X-User": auth.getEmail(), // get user email from auth service
```
@jnu i couldn't get the queries to work without this header. Is this the appropriate way to set this or is there a different method?

jnu replied:
hm, queries are working for me without it. this header will be **completely ignored** outside of development mode. so i think we should remove this authLink entirely (since auth is handled by the brother with cookies) ... it's possible it's doing some unintentional meddling like ignoring cookies.

since there's a lot in flight it might be hard to debug what the issue might be. i suggest:
1) remove the authlink, so we don't have a workaround in place that only works in dev but not prod (that'll be harder to debug)
2) merge your PR and my #74 into main
3) we both pull latest main 
4) we both reset our local development databases
5) see if there's still an issue and debug as necessary

how does that sound?

seerocode replied:
OK I removed it and rebased changes from main since #74 was merged in.

jnu replied:
> since auth is handled by the brother with cookies

i think that meant to see _browser_ ... 😅

jnu replied:
see = say 

i'm gonna stop typing ...

seerocode replied:
I like brother better lol

jnu replied:
![giphy](https://user-images.githubusercontent.com/1069899/122430305-7a908100-cf61-11eb-907d-fb11e7de8712.gif)

jnu commented on api/app_test.py:
```diff
@@ -802,7 +802,7 @@ def test_create_record(self):
                                 "name": "Trans women"
                             },
-                            "inputter": {"id": "cd7e6d44-4b4d-4d7a-8a67-31efffe53e77"}
+                            "inputter": {"id": result["data"]["createRecord"]["entries"][0]["inputter"]["id"]}
```
can you use `user.id` here? that way it'll assert that the code that infers the inputter automatically is working correctly (as written `inputter['id']` could be anything because we're just comparing it to itself)

seerocode replied:
Sure thing! It makes sense to have done this in the first place 🤦🏽

jnu commented on api/database.py:
```diff
@@ -213,7 +213,7 @@ def capitalize_tag_name(self, key, name):
 
 
-class Target(Base):
+class Target(Base, PermissionsMixin):
```
thanks for adding this!

jnu commented on client/src/services/auth.ts:
```diff
@@ -68,6 +68,28 @@ export class Auth {
 
+  /**
+   * The UUID of the currently authenticated user.
+   * @returns {string} UUID
```
i'm always on the fence about writing documentation for simple functions in typescript. docs are hard to keep in sync with code, and typescript is already able to tell us that the return type is a string.

jnu reviewed (approved):
this looks good! i wrote a couple things that i think we should address before merging:
1) the python test i think needs to use `user.id`
2) the `x-user` header makes me nervous to keep in, since by design it won't work in prod, so we need to tackle that issue head-on now (my suggestion is to remove that authLink entirely and we can see if the issue still happens once we merge everything in to `main`)

great work!!!

seerocode committed:
test: set max diff to none in setup

seerocode committed:
refactor: remove auth link for prod

seerocode committed:
test: update test to use user.id for inputter
